### ISSUE TITLE
Provide a place for user-specific configuration and files

### ISSUE THREAD

**Issue by irwir** [Mbed-TLS/mbedtls#9869]
### Suggested enhancement
Add a directory (at the library root) for user-specific settings and files - for example, `custom`.
The directory contains a number of empty files as placeholders. There should be:
`custom/mbedtls_user_config.h`,
`custom/crypto_user_config.h`,
`custom/threading_alt.h`
... all other files corresponding to '_ALT' macros.

### Justification
Users should not edit `mbedtls_config.h`; all settings go to `custom/mbedtls_user_config.h` as `#define` / `#undef` directives.
The same applies to `crypto_config.h`
Of course, both `custom/*_user_config.h` have to be included unconditionally (could be the last line of the corresponding `*_config.h`).

Mbed TLS needs this because 
library code could be treated as "read-only", user-alterable files have a separate and predefined location.
"Include" paths in projects need no modifications; simply put `#include` directive into the provided placeholder files.
This should simplify merging, updates and maintenance.
Currently,
- `MBEDTLS_USER_CONFIG_FILE` is commented out (points to null device), therefore users have to alter `mbedtls_config.h`;
- "*_alt.h" have no specific path, hence either project has to be edited, or library and user files would make a mess in the same folders.

**Comment by mpg** [comment]
Thanks for the suggestion.

We had plans about splitting the configuration file in a somewhat similar way. Our goal was to get a better separation between "user config" and "list of all tunable options for reference", because the confusion between the two causes issues in some of our test scripts. But the end result would be similar: better separation between a "read-only" part and an "editable" part. 

This kind of change can only be done in a major version, because it's a breaking API change. Unfortunately, at this point it seems unlikely we'll have time to do it in the upcoming one (Mbed TLS 4.0 + TF-PSA-Crypto 1.0).

Cc @gilles-peskine-arm

**Comment by gilles-peskine-arm** [comment]
A related proposal: https://github.com/Mbed-TLS/mbedtls/issues/9664

**Comment by irwir** [comment]
> This kind of change can only be done in a major version, because it's a breaking API change.

Where is the break?
User still could ignore the news and continue editing library files the old way.
File changes are limited to 2 `#define`s in `*_config.h` files - uncomment `*_USER_CONFIG_FILE` and give it a standard value.
The default header files are empty and thus cannot break anything even when included unconditionally.

Also, @gilles-peskine-arm suggested to backport to 3.6 branch in his issue; this hints that the changes are minor.
No real need to wait for 5.0 release.

Additionally, OpenSSL could be mentioned here - before building it, user had to run a script that created a file with the desired settings. This file does roughly the same as should do `*_user_config.h` files.

**Comment by gilles-peskine-arm** [comment]
> Users should not edit mbedtls_config.h; all settings go to custom/mbedtls_user_config.h as #define / #undef directives.

I'm not sure I understand. How does that differ from defining `MBEDTLS_CONFIG_FILE` to `"custom/mbedtls_user_config.h"`?

> Of course, both custom/*_user_config.h have to be included unconditionally (could be the last line of the corresponding *_config.h).

I don't understand how that could work. What we distribute has to work for different kinds of users, including:

* Users who want to build the library unmodified. They don't create or edit any file, and they don't define any particular identifier.
* Users who don't want to edit the source tree at all, but want to fully customize the configuration, and have no control over the include path order. They need the ability to set `MBEDTLS_CONFIG_FILE` to a different name.
* Users who want to follow the default configuration with tweaks. They define `MBEDTLS_USER_CONFIG_FILE` and place their tweaks there.
* Users who edit the configuration through an IDE that toggles boolean options on and off by (un)commenting `#define` lines. They edit `mbedtls/mbedtls_config.h` through that IDE.

(These are not theoretical concerns: all of these scenarios have come up in the past.)

Providing empty files as placeholders would break some of these scenarios. We'd have to provide a way to override the names of these files. But then what's the advantage over the current organization?

**Comment by gilles-peskine-arm** [comment]
> This kind of change can only be done in a major version, because it's a breaking API change.

Indeed, if we stop honoring `MBEDTLS_CONFIG_FILE` and `MBEDTLS_USER_CONFIG_FILE`, it's a breaking change.

If we define `MBEDTLS_USER_CONFIG_FILE` with a default name, let's say `custom/mbedtls_user_config.h`, and we provide an empty file of that name, it's only a breaking change if there are users who are already using that name, and whose `custom/mbedtls_user_config.h` happens to be later on the include path than ours. We could avoid this break by picking a name that nobody has used, but that would have to mean that the name is unintuitive, by construction. (A randomly generated name?)

> Also, @gilles-peskine-arm suggested to backport to 3.6 branch in his issue; this hints that the changes are minor.

The change I proposed in https://github.com/Mbed-TLS/mbedtls/issues/9664 would result in releases with identical `mbedtls_config.h`. It would only affect direct use of unreleased branches from Git.

**Comment by irwir** [comment]
> I'm not sure I understand. How does that differ from defining `MBEDTLS_CONFIG_FILE` to `"custom/mbedtls_user_config.h"`?

Here is a [mock-up](https://github.com/irwir/mbedtls/tree/mbedtls_alt); framework and PSA omitted so far.
The changed library could be used in the same way as before (edit `mbed_config.h` and so on).
Also, there is no need to rush with major script updates, rearrange your test settings etc. - this could be done later.
But implementation of the change already would open possibility to begin using the new features.

Unfortunately, a small change like this now needs three times more PRs than before the repository split.

>  * Users who edit the configuration through an IDE that toggles boolean options on and off by (un)commenting `#define` lines. They edit `mbedtls/mbedtls_config.h` through that IDE.

By the way, how this case was taken into account when splitting the repository?
Now there are at least two files to be edited, and most certainly this is a major breaking change.

> Providing empty files as placeholders would break some of these scenarios.

First of all, @mpg wrote "API break". Is there any? In my case only the library got changes, other codes remained unmodified.

> The change I proposed in #9664 would result in releases with identical `mbedtls_config.h`. It would only affect direct use of unreleased branches from Git.

Indeed, your perspective is of a library developer/maintainer, while my request targets library users.

> If we define `MBEDTLS_USER_CONFIG_FILE` with a default name, let's say `custom/mbedtls_user_config.h`, and we provide an empty file of that name, it's only a breaking change if there are users who are already using that name, and whose `custom/mbedtls_user_config.h` happens to be later on the include path than ours. We could avoid this break by picking a name that nobody has used, but that would have to mean that the name is unintuitive, by construction. (A randomly generated name?)

A good point, thanks. The opening message might be updated accordingly.
The uniqueness issue could be addressed by changing rather general `custom` to `mbedtls_alt` and use of "../.." path instead of "../../mbedtls_alt" in the project; and directives in the library should add path prefix - for example: `#include "mbedtls_alt/timing_alt.h"`.
Redirection of header files could be seen in the [mock-up](https://github.com/irwir/mbedtls/tree/mbedtls_alt).
